File: pkg-r/tests/testthat/test-remote-hgvs.R
# Offline tests for hgvs remote mode. Existence is checked against the Mutalyzer
# normalizer; the biogate.remote_transport option replaces the network.

.stub_mutalyzer_status <- function(status) {
  function(url, timeout) list(status = status, body = NULL)
}

test_that("safe_ident replaces filename-unsafe characters", {
  expect_identical(
    .safe_ident("NM_004006.2:c.4375C>T"),
    "NM_004006.2_c.4375C_T"
  )
  expect_identical(.safe_ident("A/B C"), "A_B_C")
  expect_identical(.safe_ident("NM_004006.2"), "NM_004006.2")
})

test_that("the remote cache path for a variant has no unsafe characters", {
  path <- .remote_cache_path("mutalyzer", "normalize", "NM_004006.2:c.4375C>T")
  base <- basename(path)
  expect_false(grepl("[>:]", base))
  expect_identical(base, "NM_004006.2_c.4375C_T.json")
})

test_that("a valid variant is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_mutalyzer_status(200))
  res <- check_id("NM_004006.2:c.4375C>T", "hgvs", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "NM_004006.2:c.4375C>T")
  expect_true(is.na(res$suggestion))
})

test_that("a reference-inconsistent variant is invalid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_mutalyzer_status(422))
  res <- check_id("NM_004006.2:c.4375A>T", "hgvs", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_true(is.na(res$suggestion))
})

test_that("a malformed variant skips the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("a malformed variant must not reach the network")
    }
  )
  res <- check_id("NM_004006.2:c.76insG", "hgvs", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_mutalyzer_status(500))
  expect_error(
    check_id("NM_004006.2:c.4375C>T", "hgvs", how = "remote"),
    class = "biogate_error_remote"
  )
  expect_false(file.exists(
    .remote_cache_path("mutalyzer", "normalize", "NM_004006.2:c.4375C>T")
  ))
})

test_that("the disk cache short-circuits the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_mutalyzer_status(200))
  expect_true(check_id("NM_004006.2:c.4375C>T", "hgvs", how = "remote")$valid)
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("network should not be called on a cache hit")
    }
  )
  expect_true(check_id("NM_004006.2:c.4375C>T", "hgvs", how = "remote")$valid)
})

File: pkg-py/tests/test_remote_hgvs.py
"""HGVS remote mode: existence via the Mutalyzer normalizer, network stubbed."""

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub_status(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_safe_ident_replaces_unsafe_characters():
    assert remote._safe_ident("NM_004006.2:c.4375C>T") == "NM_004006.2_c.4375C_T"
    assert remote._safe_ident("A/B C") == "A_B_C"
    # a plain accession keeps its dot and underscore
    assert remote._safe_ident("NM_004006.2") == "NM_004006.2"


def test_cache_path_has_no_unsafe_characters():
    path = remote._remote_cache_path("mutalyzer", "normalize", "NM_004006.2:c.4375C>T")
    assert ">" not in path.name and ":" not in path.name
    assert path.name == "NM_004006.2_c.4375C_T.json"


def test_valid_variant_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_status(200))
    res = biobouncer.check_id("NM_004006.2:c.4375C>T", source_db="hgvs", how="remote")[
        0
    ]
    assert res.valid is True
    assert res.normalized == "NM_004006.2:c.4375C>T"
    assert res.suggestion is None
    assert res.how == "remote"


def test_reference_inconsistent_variant_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_status(422))
    res = biobouncer.check_id("NM_004006.2:c.4375A>T", source_db="hgvs", how="remote")[
        0
    ]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_malformed_variant_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed variant must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    # insertion without a flanking range fails the offline grammar first
    res = biobouncer.check_id("NM_004006.2:c.76insG", source_db="hgvs", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_status(500))
    with pytest.raises(RemoteError):
        biobouncer.check_id("NM_004006.2:c.4375C>T", source_db="hgvs", how="remote")
    path = remote._remote_cache_path("mutalyzer", "normalize", "NM_004006.2:c.4375C>T")
    assert not path.is_file()


def test_cache_round_trip(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_status(200))
    assert biobouncer.check_id("NM_004006.2:c.4375C>T", source_db="hgvs", how="remote")[
        0
    ].valid

    def _forbidden(url, timeout=30):
        raise AssertionError("network should not be called on a cache hit")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    assert biobouncer.check_id("NM_004006.2:c.4375C>T", source_db="hgvs", how="remote")[
        0
    ].valid
